{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "2b36998d-47c4-4447-95f6-a1d21f02dc1c",
   "metadata": {},
   "source": [
    "# 🦠 Building your own Genome Scale Model with ChiMera\n",
    "\n",
    "\n",
    "<img src=\"input_examples/assets/ChiMera_logo.png\" alt=\"ChiMera logo\" style=\"height: 200px; width:200px;margin-left: 330px;\"/>\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "This notebook will teach the basics of ChiMera.\n",
    "\n",
    "## About the Install and Data\n",
    "\n",
    "The folder `input_examples` will have all necessary data to perform the Tutorial!\n",
    "\n",
    "If you don't have ChiMera, you can Download via our [GitHub](https://github.com/tamascogustavo/chimera) or [PyPI](https://pypi.org/project/ChiMera-ModelBuilder/).\n",
    "\n",
    "Remember you need also to install [CPLEX](https://community.ibm.com/community/user/datascience/blogs/xavier-nodet1/2020/07/09/cplex-free-for-students). If you find problems during the installation, check our GitHub.\n",
    "\n",
    "## Special thanks to \n",
    "\n",
    "\n",
    " + Predefined maps generation by Escher\n",
    " \n",
    "    King, Z.A., Dräger, A., Ebrahim, A., Sonnenschein, N., Lewis, N.E. and Palsson, B.O., 2015. Escher: a web application for building, sharing, and embedding data-rich visualizations of biological pathways. PLoS computational biology, 11(8), p.e1004321.\n",
    "    \n",
    "+ Carving process by CarveMe\n",
    "\n",
    "    D. Machado et al, \"Fast automated reconstruction of genome-scale metabolic models for microbial species and communities\", Nucleic Acids Research, gky537, 2018. doi: https://doi.org/10.1093/nar/gky537   \n",
    "+ Model manipulation by Cobrapy\n",
    "\n",
    "    Ebrahim, A., Lerman, J.A., Palsson, B.O. and Hyduke, D.R., 2013. COBRApy: constraints-based reconstruction and analysis for python. BMC systems biology, 7(1), pp.1-6.\n",
    "    \n",
    "+ Generation of Cytoscape compatible maps with PSAMMr\n",
    "    Steffensen, J.L., Dufault-Thompson, K. and Zhang, Y., 2016. PSAMM: a portable system for the analysis of metabolic models. PLoS computational bio\n",
    "    \n",
    "## Don’t forget to cite them !!! And us as well 😃\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16caadd8-fda4-43ad-a98e-fc5e68871a55",
   "metadata": {},
   "source": [
    "# The basics of ChiMera"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8724aa34-d286-47e5-b963-8a2649d5433e",
   "metadata": {},
   "source": [
    "The command `python chimera_core.py -h` should prompt a short manual about ChiMera usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e1791d07-2051-49cf-859b-d02884b80f3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: chimera_core.py [-h] {core,silencing,harvest_path_cytoscape} ...\n",
      "\n",
      "positional arguments:\n",
      "  {core,silencing,harvest_path_cytoscape}\n",
      "    core                Runs the core module\n",
      "    silencing           Runs the knockout module\n",
      "    harvest_path_cytoscape\n",
      "                        Adds kegg pathway information to cytoscape maps\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "458e8b21-7307-494d-b7e0-a06a32eb1367",
   "metadata": {},
   "source": [
    "## Here you can see all the modules of Chimera:\n",
    "\n",
    "To check the requirements of each of the modules, you can perform the same task again, just by adding the name of the module this time.\n",
    "\n",
    "The output will explain the requirements to perform the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "76476fdf-eca5-415d-b134-a211adbce3cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: chimera_core.py core [-h] --organism ORGANISM --type TYPE --media MEDIA\n",
      "                            [--mediadb MEDIADB]\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help           show this help message and exit\n",
      "  --organism ORGANISM  path to the *.faa file\n",
      "  --type TYPE          type of organism, options are gramneg, grampos,\n",
      "                       bacteria\n",
      "  --media MEDIA        Growth media ID\n",
      "  --mediadb MEDIADB    tsv file with a new media composition for new\n",
      "                       experimental conditions. If provided you must pass the\n",
      "                       media id in --media parameter.\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py core -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "eac54d09-5c57-4ca6-b31d-50bd4e7dc2f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: chimera_core.py silencing [-h] --i I --type TYPE --targets TARGETS\n",
      "                                 --faa FAA --mode MODE\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help         show this help message and exit\n",
      "  --i I              path to the *.sbml model file\n",
      "  --type TYPE        type of knockout target, gene or reaction\n",
      "  --targets TARGETS  path to csv file containing targets, one by line\n",
      "  --faa FAA          path to the faa file\n",
      "  --mode MODE        Type of knockout: single or douple or all. For double all\n",
      "                     combinations of targets will be performed\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py silencing -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "59f33c3a-cde0-4f60-aa4a-f1a28211ddc7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: chimera_core.py harvest_path_cytoscape [-h] --table TABLE --model MODEL\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help     show this help message and exit\n",
      "  --table TABLE  Path to reactions.edges.tsv file inside psam_* folder\n",
      "  --model MODEL  path to sbml model file\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py harvest_path_cytoscape -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f0ea84c-2de9-44aa-b845-046b56c5ab26",
   "metadata": {},
   "source": [
    "## Let's us start with the CORE module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "604a284e-e972-4ff7-99a5-8174a83a56ad",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------------------------------------\n",
      "Building Model\n",
      "---------------------------------------------------\n",
      "diamond v2.0.9.147 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "Please cite: http://dx.doi.org/10.1038/s41592-021-01101-x Nature Methods (2021)\n",
      "\n",
      "#CPU threads: 12\n",
      "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
      "Temporary directory: input_examples/faa_file\n",
      "Percentage range of top alignment score to report hits: 10\n",
      "Opening the database...  [0.013s]\n",
      "Database: /Users/gustavotamasco/opt/anaconda3/lib/python3.7/site-packages/carveme/data/generated/bigg_proteins.dmnd (type: Diamond database, sequences: 26727, letters: 11170577)\n",
      "Block size = 2000000000\n",
      "Opening the input file...  [0.002s]\n",
      "Opening the output file...  [0s]\n",
      "Loading query sequences...  [0.011s]\n",
      "Masking queries...  [0.018s]\n",
      "Algorithm: Double-indexed\n",
      "Building query histograms...  [0.03s]\n",
      "Allocating buffers...  [0s]\n",
      "Loading reference sequences...  [0.013s]\n",
      "Masking reference...  [0.081s]\n",
      "Initializing temporary storage...  [0.001s]\n",
      "Building reference histograms...  [0.198s]\n",
      "Allocating buffers...  [0s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.066s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.037s]\n",
      "Building query seed array...  [0.007s]\n",
      "Computing hash join...  [0.007s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.039s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.029s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 1/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.019s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.024s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.033s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.025s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.03s]\n",
      "Processing query block 1, reference block 1/1, shape 2/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.021s]\n",
      "Building query seed array...  [0.002s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.03s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.021s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.037s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.005s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.036s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.005s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 3/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.022s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.024s]\n",
      "Building query seed array...  [0.005s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.038s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.029s]\n",
      "Building query seed array...  [0.005s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.038s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 4/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.021s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.031s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.022s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.036s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.025s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.031s]\n",
      "Processing query block 1, reference block 1/1, shape 5/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.023s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.022s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.04s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.005s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.035s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.033s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 6/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.025s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.035s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.038s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.033s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.03s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 7/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.022s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.031s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.024s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.036s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.029s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.039s]\n",
      "Processing query block 1, reference block 1/1, shape 8/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.029s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.036s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.04s]\n",
      "Building query seed array...  [0.005s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.04s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.027s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.035s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.029s]\n",
      "Building query seed array...  [0.006s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 9/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.021s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.001s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.024s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.036s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.027s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.033s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.034s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 10/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.024s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.027s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.039s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.025s]\n",
      "Building query seed array...  [0.006s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 11/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.025s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.003s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.037s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.03s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.035s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.027s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 12/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.023s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.033s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.023s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.038s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.027s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.029s]\n",
      "Building query seed array...  [0.006s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 13/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.024s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.003s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.024s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.003s]\n",
      "Searching alignments...  [0.037s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.033s]\n",
      "Processing query block 1, reference block 1/1, shape 14/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.023s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.032s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.023s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.037s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.026s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.003s]\n",
      "Searching alignments...  [0.035s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.037s]\n",
      "Processing query block 1, reference block 1/1, shape 15/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.028s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.034s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 1/4.\n",
      "Building reference seed array...  [0.027s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.039s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 2/4.\n",
      "Building reference seed array...  [0.031s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.004s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.035s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 3/4.\n",
      "Building reference seed array...  [0.033s]\n",
      "Building query seed array...  [0.004s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.036s]\n",
      "Processing query block 1, reference block 1/1, shape 16/16, index chunk 4/4.\n",
      "Building reference seed array...  [0.023s]\n",
      "Building query seed array...  [0.003s]\n",
      "Computing hash join...  [0.005s]\n",
      "Building seed filter...  [0.002s]\n",
      "Searching alignments...  [0.032s]\n",
      "Deallocating buffers...  [0.001s]\n",
      "Clearing query masking...  [0s]\n",
      "Computing alignments...  [0.697s]\n",
      "Deallocating reference...  [0s]\n",
      "Loading reference sequences...  [0s]\n",
      "Deallocating buffers...  [0s]\n",
      "Deallocating queries...  [0s]\n",
      "Loading query sequences...  [0s]\n",
      "Closing the input file...  [0s]\n",
      "Closing the output file...  [0.002s]\n",
      "Closing the database file...  [0.002s]\n",
      "Deallocating taxonomy...  [0s]\n",
      "Total time = 5.832s\n",
      "Reported 9257 pairwise alignments, 9257 HSPs.\n",
      "2275 queries aligned.\n",
      "Exit status: 0\n",
      "e_coli_test.sbml SBML model was created\n",
      "Basic info of the model\n",
      "-----------------------\n",
      "model has 2665 reactions\n",
      "model has 1668 metabolites\n",
      "model has 1702 genes\n",
      "The default flux bounds are:(-1000.0, 1000.0)\n",
      "Maximize\n",
      "1.0*Growth - 1.0*Growth_reverse_699ae\n",
      "max\n",
      "objective expression 1.0*Growth - 1.0*Growth_reverse_699ae\n",
      "<Solution 0.674 at 0x7f91cad01d10>\n",
      "Objective\n",
      "=========\n",
      "1.0 Growth = 0.674350736633079\n",
      "\n",
      "Uptake\n",
      "------\n",
      " Metabolite       Reaction      Flux  C-Number  C-Flux\n",
      "    M_ca2_e     R_EX_ca2_e   0.00351         0   0.00%\n",
      "     M_cl_e      R_EX_cl_e   0.00351         0   0.00%\n",
      "M_cobalt2_e R_EX_cobalt2_e 6.744E-05         0   0.00%\n",
      "    M_cu2_e     R_EX_cu2_e 0.0004781         0   0.00%\n",
      "    M_fe2_e     R_EX_fe2_e  0.004829         0   0.00%\n",
      "    M_fe3_e     R_EX_fe3_e  0.005265         0   0.00%\n",
      " M_glc__D_e  R_EX_glc__D_e        10         6 100.00%\n",
      "      M_k_e       R_EX_k_e    0.1316         0   0.00%\n",
      "    M_mg2_e     R_EX_mg2_e   0.00585         0   0.00%\n",
      "    M_mn2_e     R_EX_mn2_e  0.000466         0   0.00%\n",
      "    M_nh4_e     R_EX_nh4_e     7.273         0   0.00%\n",
      "     M_o2_e      R_EX_o2_e        10         0   0.00%\n",
      "     M_pi_e      R_EX_pi_e    0.6501         0   0.00%\n",
      "    M_so4_e     R_EX_so4_e     0.169         0   0.00%\n",
      "    M_zn2_e     R_EX_zn2_e   0.00023         0   0.00%\n",
      "\n",
      "Secretion\n",
      "---------\n",
      "Metabolite    Reaction       Flux  C-Number C-Flux\n",
      "  M_4hba_e R_EX_4hba_e -0.0001504         7  0.00%\n",
      "   M_co2_e  R_EX_co2_e      -18.4         1 56.92%\n",
      "  M_etoh_e R_EX_etoh_e     -6.964         2 43.08%\n",
      "   M_h2o_e  R_EX_h2o_e     -29.24         0  0.00%\n",
      "     M_h_e    R_EX_h_e     -6.201         0  0.00%\n",
      "\n",
      "e_coli_test.json was created.\n",
      "---------------------------------------------------\n",
      "Generating Escher Metabolic Maps\n",
      "---------------------------------------------------\n",
      "formatted_e_coli_test_tryptophan_metabolism.html was created\n",
      "formatted_e_coli_test_central_metabolism.html was created\n",
      "formatted_e_coli_test_carbohydrate_metabolism.html was created\n",
      "formatted_e_coli_test_central_carbon_metabolism_final.html was created\n",
      "formatted_e_coli_test_nucleotide_metabolism.html was created\n",
      "formatted_e_coli_test_fatty_acid_biosynthesis_saturated.html was created\n",
      "formatted_e_coli_test_fatty_acid_beta_oxidation.html was created\n",
      "formatted_e_coli_test_partial_amino_acid metabolism.html was created\n",
      "formatted_e_coli_test_glycolysis_TCA_PPP.html was created\n",
      "formatted_e_coli_test_inositol_retinol_metabolism.html was created\n",
      "<<< All escher maps are in : /Users/gustavotamasco/prometeus/gsmr/chimera/metabolism_maps>>>\n",
      "---------------------------------------------------\n",
      "Creating Cytoscape Maps\n",
      "---------------------------------------------------\n",
      "psamm_e_coli_test\n",
      "WARNING: Unable to parse charge for M_1agpg161_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_1p2cbxl_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_23ddhb_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_23dhbzs_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2agpg120_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2agpg120_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2agpg180_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2agpg180_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2ahbut_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2dr1p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_2hdecg3p_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2hdecg3p_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_2me4p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_2shchc_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_3hocoa_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_3hodcoa_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_3uib_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_5aizc_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_5hiu_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_6p2dhglcn_c as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_6pgg_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_acgam1p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_acgam1p_e as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_acgam1p_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_acmanap_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_acmum6p_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_air_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_anhgm3p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_anhgm3p_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_argsuc_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_ascb__L_e as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_ascb__L_p as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_crncoa_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_ctbtcoa_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_db4p_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_dcamp_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_dhgly_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_dhpmp_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_dmlz_c as an integer: 1.0\n",
      "WARNING: Unable to parse charge for M_dscl_c as an integer: -7.0\n",
      "WARNING: Unable to parse charge for M_fadh2_c as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_fc1p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_fdp_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_fe3dcit_c as an integer: -12.0\n",
      "WARNING: Unable to parse charge for M_fe3dcit_e as an integer: -12.0\n",
      "WARNING: Unable to parse charge for M_fe3dcit_p as an integer: -12.0\n",
      "WARNING: Unable to parse charge for M_fe3hox_un_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_fe3hox_un_e as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_fe3hox_un_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_feenter_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_feenter_e as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_feenter_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_feoxam_un_c as an integer: 1.0\n",
      "WARNING: Unable to parse charge for M_feoxam_un_e as an integer: 1.0\n",
      "WARNING: Unable to parse charge for M_feoxam_un_p as an integer: 1.0\n",
      "WARNING: Unable to parse charge for M_fgam_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_fmnh2_c as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_fpram_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_frulysp_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_fruur_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_fruur_e as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_fruur_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_g3p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_g3pg_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_g3pg_e as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_g3pg_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_gdptp_c as an integer: -7.0\n",
      "WARNING: Unable to parse charge for M_glcur1p_e as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_glcur1p_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_h2s_c as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_h2s_e as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_h2s_p as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_hkntd_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_man1p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_man6p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_man6p_e as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_man6p_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_man6pglyc_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_murein3px3p_p as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_murein4px4p_p as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_murein5px4p_p as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_murein5px4px4p_p as an integer: -6.0\n",
      "WARNING: Unable to parse charge for M_n2o_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_n2o_e as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_n2o_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_oc2coa_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_op4en_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_pa120_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa120_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa140_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa140_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa141_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa141_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa161_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa161_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa180_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa180_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa181_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pa181_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pg120_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_pg120_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_pg160_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_pg160_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_pg180_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_pg180_p as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_pgp120_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp120_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp140_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp140_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp141_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp141_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp160_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp160_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp161_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp161_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp180_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp180_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp181_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_pgp181_p as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_ppgpp_c as an integer: -6.0\n",
      "WARNING: Unable to parse charge for M_pphn_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_pppi_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_pprdn_c as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_pqq_p as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_prbamp_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_prbatp_c as an integer: -6.0\n",
      "WARNING: Unable to parse charge for M_ps160_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_ps161_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_ps181_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_r5p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_r5p_e as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_r5p_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_rbflvrd_c as an integer: 2.0\n",
      "WARNING: Unable to parse charge for M_ribflv_c as an integer: 2.0\n",
      "WARNING: Unable to parse charge for M_ribflv_e as an integer: 2.0\n",
      "WARNING: Unable to parse charge for M_rml1p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_sbt6p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_sbzcoa_c as an integer: -5.0\n",
      "WARNING: Unable to parse charge for M_scl_c as an integer: -7.0\n",
      "WARNING: Unable to parse charge for M_seramp_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_ssaltpp_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_suc6p_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_tag6p__D_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_tagdp__D_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_tagur_c as an integer: -1.0\n",
      "WARNING: Unable to parse charge for M_td2coa_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_tdecoa_c as an integer: -4.0\n",
      "WARNING: Unable to parse charge for M_tgt_e as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_tsul_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_tsul_e as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_tsul_p as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_tton_e as an integer: 0.0\n",
      "WARNING: Unable to parse charge for M_uaagmda_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_uagmda_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_uamr_c as an integer: -3.0\n",
      "WARNING: Unable to parse charge for M_udcpdp_c as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_udpacgal_e as an integer: -2.0\n",
      "WARNING: Unable to parse charge for M_udpacgal_p as an integer: -2.0\n",
      "INFO: e is extracellular compartment\n",
      "Model: None\n",
      "- Biomass reaction: Growth\n",
      "- Compartments: 3\n",
      "- Compounds: 1668\n",
      "- Reactions: 2319\n",
      "- Genes: 0\n",
      "INFO: Using default flux limit of 1000\n",
      "INFO: Converting exchange reactions to exchange file\n",
      "Exit status: 0\n",
      "psamm_e_coli_test was created\n",
      "INFO:numexpr.utils:Note: NumExpr detected 12 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO: Model version: b'54c3428-dirty'\n",
      "WARNING: The compound cd2_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound cm_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound cro4_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound doxrbcn_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound fusa_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound gbbtn_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound hg2_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound na1_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound ni2_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound novbcn_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: The compound ttrcyc_e was in the extracellular compartment but not defined in the exchange compounds\n",
      "WARNING: Reaction Growth is in the subset and exclude file. Reaction will be excluded.\n",
      "INFO: Iteration 1: 2314 reactions...\n",
      "INFO: Iteration 2: 2314 reactions...\n",
      "INFO: Iteration 3: 950 reactions...\n",
      "INFO: Iteration 4: 20 reactions...\n",
      "The status is 0\n",
      "The .tsv files for cytoscape were generated\n",
      "Harvesting metabolic pathways from model. This may take a while...\n",
      "\u001b[2m\u001b[36m(get_compound_info pid=37636)\u001b[0m /Users/gustavotamasco/opt/anaconda3/lib/python3.7/site-packages/bs4/__init__.py:439: MarkupResemblesLocatorWarning: The input looks more like a filename than markup. You may want to open this file and pass the filehandle into Beautiful Soup.\n",
      "\u001b[2m\u001b[36m(get_compound_info pid=37636)\u001b[0m   MarkupResemblesLocatorWarning\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py core --organism input_examples/faa_file/e_coli_test.faa --type gramneg --media M9"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dcbf003-952f-483b-b1e0-55d62325a20c",
   "metadata": {},
   "source": [
    "\n",
    "***As you can see, a lot of data is produced.***\n",
    "\n",
    "### Let's go over the results:\n",
    "\n",
    "First, we can see that a model is created and loaded into Cobrapy. There we perform FBA using the defined biomass equation.\n",
    "\n",
    "```\n",
    "e_coli_test.sbml SBML model was created\n",
    "Basic info of the model\n",
    "-----------------------\n",
    "model has 2665 reactions\n",
    "model has 1668 metabolites\n",
    "model has 1702 genes\n",
    "\n",
    "```\n",
    "\n",
    "Next is prompted to the user the table of predicted growth rate and fluxes.\n",
    "Don't worry, we also save the output in a CSV file, so you can check the data later (the extension is `_uptake_secretion_rates.csv`).\n",
    "\n",
    "```\n",
    "\n",
    "<Solution 0.674 at 0x7f91cad01d10>\n",
    "Objective\n",
    "=========\n",
    "1.0 Growth = 0.674350736633079\n",
    "\n",
    "Uptake\n",
    "------\n",
    " Metabolite       Reaction      Flux  C-Number  C-Flux\n",
    "    M_ca2_e     R_EX_ca2_e   0.00351         0   0.00%\n",
    "     M_cl_e      R_EX_cl_e   0.00351         0   0.00%\n",
    "M_cobalt2_e R_EX_cobalt2_e 6.744E-05         0   0.00%\n",
    "    M_cu2_e     R_EX_cu2_e 0.0004781         0   0.00%\n",
    "    M_fe2_e     R_EX_fe2_e  0.004829         0   0.00%\n",
    "    M_fe3_e     R_EX_fe3_e  0.005265         0   0.00%\n",
    " M_glc__D_e  R_EX_glc__D_e        10         6 100.00%\n",
    "      M_k_e       R_EX_k_e    0.1316         0   0.00%\n",
    "    M_mg2_e     R_EX_mg2_e   0.00585         0   0.00%\n",
    "    M_mn2_e     R_EX_mn2_e  0.000466         0   0.00%\n",
    "    M_nh4_e     R_EX_nh4_e     7.273         0   0.00%\n",
    "     M_o2_e      R_EX_o2_e        10         0   0.00%\n",
    "     M_pi_e      R_EX_pi_e    0.6501         0   0.00%\n",
    "    M_so4_e     R_EX_so4_e     0.169         0   0.00%\n",
    "    M_zn2_e     R_EX_zn2_e   0.00023         0   0.00%\n",
    "\n",
    "Secretion\n",
    "---------\n",
    "Metabolite    Reaction       Flux  C-Number C-Flux\n",
    "  M_4hba_e R_EX_4hba_e -0.0001504         7  0.00%\n",
    "   M_co2_e  R_EX_co2_e      -18.4         1 56.92%\n",
    "  M_etoh_e R_EX_etoh_e     -6.964         2 43.08%\n",
    "   M_h2o_e  R_EX_h2o_e     -29.24         0  0.00%\n",
    "     M_h_e    R_EX_h_e     -6.201         0  0.00%\n",
    "\n",
    "```\n",
    "\n",
    "Next, we convert the model to different formats, allowing users to load the produced model in different software. \n",
    "\n",
    "We also use the model to perform metabolic pathway detection. The results are located in the folder **metabolism_maps**.  There you will find the maps of the most important pathways. Data in blue indicates presence in your model. Red indicates absence.\n",
    "\n",
    "Inside the **psamm folder**, you will find YAML files, explaining the model composition. With them, you can check the reactions, compounds, and more. The `reactions.edges` and `reactions.nodes` are compatible with Gephi and Cytoscape, so you can load your graphs in this software to visualize you model. In our GitHub, you have links explaining how to do that.\n",
    "\n",
    "We also produce a CSV file that contains the `enriched detected pathways` in your model. We use KEGG API for that. The same file is used to produce an `HTML interactive image` of the 30 most detected pathways. \n",
    "\n",
    "\n",
    "![ChiMera Paths](input_examples/assets/chimera_paths.png \"Top 30 detected Pathways\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "026e90a3-4f25-4901-9b08-6e3244f4bd9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0089b499-ff76-4e61-8147-996a305bc81e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Pathway</th>\n",
       "      <th>Counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>map00051  Fructose and mannose metabolism</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>map01100  Metabolic pathways</td>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>map02010  ABC transporters</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>map02060  Phosphotransferase system (PTS)</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>map00270  Cysteine and methionine metabolism</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>map00670  One carbon pool by folate</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>map00720  Carbon fixation pathways in prokaryotes</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>map00970  Aminoacyl-tRNA biosynthesis</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>map01120  Microbial metabolism in diverse envi...</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>map01200  Carbon metabolism</td>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Pathway  Counts\n",
       "0          map00051  Fructose and mannose metabolism      17\n",
       "1                       map01100  Metabolic pathways     321\n",
       "2                         map02010  ABC transporters      43\n",
       "3          map02060  Phosphotransferase system (PTS)      21\n",
       "4       map00270  Cysteine and methionine metabolism      18\n",
       "5                map00670  One carbon pool by folate       3\n",
       "6  map00720  Carbon fixation pathways in prokaryotes      19\n",
       "7              map00970  Aminoacyl-tRNA biosynthesis      14\n",
       "8  map01120  Microbial metabolism in diverse envi...     109\n",
       "9                        map01200  Carbon metabolism      43"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reactions = pd.read_csv(\"input_examples/treasure_chest/e_coli_test_enriched_paths.csv\")\n",
    "reactions.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "463b282f-76e3-4d95-88c6-23b0009e9506",
   "metadata": {},
   "source": [
    "## The hasvest module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3bc50e6a-4506-4abb-9d94-3a2cbd526a77",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Process initiated, this can take some time!\n",
      "735\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py harvest_path_cytoscape --table input_examples/treasure_chest/psamm_e_coli_test/reactions.edges.tsv  --model input_examples/treasure_chest/e_coli_test.sbml "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d81fcb6-9c9c-4f0c-b400-03968ad679b1",
   "metadata": {},
   "source": [
    "The command will create a new `reactions_edges.tsv` file inside the main folder. This file now have information about the pathways that the compounds participate.\n",
    "\n",
    "### Let's take a look\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1cb1d0a5-5c59-478d-9152-8a7308b3d8fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>source</th>\n",
       "      <th>target</th>\n",
       "      <th>direction</th>\n",
       "      <th>penwidth</th>\n",
       "      <th>style</th>\n",
       "      <th>source_pathway</th>\n",
       "      <th>target_pathway</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10fthf_c[c]</td>\n",
       "      <td>GARFT_1</td>\n",
       "      <td>both</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GARFT_1</td>\n",
       "      <td>fgam_c[c]</td>\n",
       "      <td>both</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>map00230  Purine metabolism | map01100  Metabo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10fthf_c[c]</td>\n",
       "      <td>FTHFD_1</td>\n",
       "      <td>forward</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FTHFD_1</td>\n",
       "      <td>for_c[c]</td>\n",
       "      <td>forward</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10fthf_c[c]</td>\n",
       "      <td>AICART_1</td>\n",
       "      <td>both</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AICART_1</td>\n",
       "      <td>fprica_c[c]</td>\n",
       "      <td>both</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>map00230  Purine metabolism | map01060  Biosyn...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AICART_1</td>\n",
       "      <td>thf_c[c]</td>\n",
       "      <td>both</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GARFT_1</td>\n",
       "      <td>thf_c[c]</td>\n",
       "      <td>both</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FTHFD_1</td>\n",
       "      <td>thf_c[c]</td>\n",
       "      <td>forward</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>12dgr120_c[c]</td>\n",
       "      <td>DAGK120_1</td>\n",
       "      <td>forward</td>\n",
       "      <td>1</td>\n",
       "      <td>solid</td>\n",
       "      <td>no pathway detected</td>\n",
       "      <td>no pathway detected</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          source       target direction penwidth  style       source_pathway  \\\n",
       "0    10fthf_c[c]      GARFT_1      both        1  solid  no pathway detected   \n",
       "1        GARFT_1    fgam_c[c]      both        1  solid  no pathway detected   \n",
       "2    10fthf_c[c]      FTHFD_1   forward        1  solid  no pathway detected   \n",
       "3        FTHFD_1     for_c[c]   forward        1  solid  no pathway detected   \n",
       "4    10fthf_c[c]     AICART_1      both        1  solid  no pathway detected   \n",
       "5       AICART_1  fprica_c[c]      both        1  solid  no pathway detected   \n",
       "6       AICART_1     thf_c[c]      both        1  solid  no pathway detected   \n",
       "7        GARFT_1     thf_c[c]      both        1  solid  no pathway detected   \n",
       "8        FTHFD_1     thf_c[c]   forward        1  solid  no pathway detected   \n",
       "9  12dgr120_c[c]    DAGK120_1   forward        1  solid  no pathway detected   \n",
       "\n",
       "                                      target_pathway  \n",
       "0                                no pathway detected  \n",
       "1  map00230  Purine metabolism | map01100  Metabo...  \n",
       "2                                no pathway detected  \n",
       "3                                no pathway detected  \n",
       "4                                no pathway detected  \n",
       "5  map00230  Purine metabolism | map01060  Biosyn...  \n",
       "6                                no pathway detected  \n",
       "7                                no pathway detected  \n",
       "8                                no pathway detected  \n",
       "9                                no pathway detected  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_table(\"input_examples/treasure_chest/e_coli_test_reactions_edges.tsv\")\n",
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19cb82dd-72ca-4799-9d0c-0abaaa54db4b",
   "metadata": {},
   "source": [
    "## Finally we have the knockout module"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ff67be5-8607-496f-9945-fb8a80609b15",
   "metadata": {},
   "source": [
    "The logic that we will show here is the same for all the possible combinations of parameters.\n",
    "\n",
    "First, we will perform an example of a single gene, after which we will make a double reaction knockout."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09286ca4-5f96-4944-8b6d-a5735f726fe7",
   "metadata": {},
   "source": [
    "### Single gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8c053671-c066-412d-bdd4-14a2704ee3fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "complete model growth rate:  <Solution 0.674 at 0x7fb9adcda810>\n",
      "\n",
      "\n",
      "The knockout effect of trkA\n",
      "                                             ids    growth   status\n",
      "282  {lcl_NZ_CP027599_1_prot_WP_000691382_1_476}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of gltP\n",
      "                                              ids    growth   status\n",
      "365  {lcl_NZ_CP027599_1_prot_WP_000835799_1_5587}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of cysQ\n",
      "                                               ids    growth   status\n",
      "1654  {lcl_NZ_CP027599_1_prot_WP_000886909_1_5394}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of argC\n",
      "                                               ids  growth   status\n",
      "1275  {lcl_NZ_CP027599_1_prot_WP_000935363_1_5705}     0.0  optimal\n",
      "\n",
      "\n",
      "The knockout effect of pgpA\n",
      "                                              ids    growth   status\n",
      "852  {lcl_NZ_CP027599_1_prot_WP_000154044_1_4715}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of edd\n",
      "                                              ids    growth   status\n",
      "334  {lcl_NZ_CP027599_1_prot_WP_001069467_1_2511}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of dctA\n",
      "                                              ids    growth   status\n",
      "1074  {lcl_NZ_CP027599_1_prot_WP_044164517_1_245}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of ddlB\n",
      "                                              ids    growth   status\n",
      "651  {lcl_NZ_CP027599_1_prot_WP_000130056_1_5076}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of thrC\n",
      "                                               ids  growth   status\n",
      "1391  {lcl_NZ_CP027599_1_prot_WP_000781063_1_5168}     0.0  optimal\n",
      "\n",
      "\n",
      "The knockout effect of map\n",
      "                                               ids    growth   status\n",
      "1025  {lcl_NZ_CP027599_1_prot_WP_001018204_1_4999}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of relA\n",
      "                                               ids    growth   status\n",
      "1342  {lcl_NZ_CP027599_1_prot_WP_000226815_1_1095}  0.674351  optimal\n",
      "\n",
      "\n",
      "The knockout effect of pyrF\n",
      "                                              ids  growth   status\n",
      "447  {lcl_NZ_CP027599_1_prot_WP_000176278_1_2958}     0.0  optimal\n",
      "\n",
      "\n",
      "The knockout effect of pyrD\n",
      "                                               ids    growth   status\n",
      "1083  {lcl_NZ_CP027599_1_prot_WP_001295352_1_4084}  0.671667  optimal\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py silencing --i input_examples/treasure_chest/e_coli_test.sbml --type gene --targets input_examples/reations_gene_to_delete/knockout_genes_list.txt --faa input_examples/faa_file/e_coli_test.faa --mode single"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fef70b12-5598-4b06-bdcd-cb71bd217095",
   "metadata": {},
   "source": [
    "### Double reaction\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "14b70281-4c09-4453-add3-740f9b616614",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "complete model growth rate:  <Solution 0.674 at 0x7fbb8be87ed0>\n",
      "\n",
      "\n",
      "The knockout effect of GLCDpp + HEX1\n",
      "<Solution 0.674 at 0x7fbb8be87bd0>\n",
      "\n",
      "\n",
      "The knockout effect of GLCDpp + GNK\n",
      "<Solution 0.674 at 0x7fbb8be87b10>\n",
      "\n",
      "\n",
      "The knockout effect of GLCDpp + PGLCNDH\n",
      "<Solution 0.674 at 0x7fbb8be87e10>\n",
      "\n",
      "\n",
      "The knockout effect of GLCDpp + EDD\n",
      "<Solution 0.674 at 0x7fbb8be87a10>\n",
      "\n",
      "\n",
      "The knockout effect of GLCDpp + EDA\n",
      "<Solution 0.674 at 0x7fbb8be87dd0>\n",
      "\n",
      "\n",
      "The knockout effect of HEX1 + GNK\n",
      "<Solution 0.674 at 0x7fbb8be87e90>\n",
      "\n",
      "\n",
      "The knockout effect of HEX1 + PGLCNDH\n",
      "<Solution 0.674 at 0x7fbb8be87b10>\n",
      "\n",
      "\n",
      "The knockout effect of HEX1 + EDD\n",
      "<Solution 0.674 at 0x7fbb8be87d90>\n",
      "\n",
      "\n",
      "The knockout effect of HEX1 + EDA\n",
      "<Solution 0.674 at 0x7fbb8be87f10>\n",
      "\n",
      "\n",
      "The knockout effect of GNK + PGLCNDH\n",
      "<Solution 0.674 at 0x7fbb8be87890>\n",
      "\n",
      "\n",
      "The knockout effect of GNK + EDD\n",
      "<Solution 0.674 at 0x7fbb8be87a10>\n",
      "\n",
      "\n",
      "The knockout effect of GNK + EDA\n",
      "<Solution 0.674 at 0x7fbb8be87950>\n",
      "\n",
      "\n",
      "The knockout effect of PGLCNDH + EDD\n",
      "<Solution 0.674 at 0x7fbb8be87b10>\n",
      "\n",
      "\n",
      "The knockout effect of PGLCNDH + EDA\n",
      "<Solution 0.674 at 0x7fbb8be87d90>\n",
      "\n",
      "\n",
      "The knockout effect of EDD + EDA\n",
      "<Solution 0.674 at 0x7fbb8be873d0>\n",
      "\u001b[0m"
     ]
    }
   ],
   "source": [
    "! chimera_core.py silencing --i input_examples/treasure_chest/e_coli_test.sbml --type reaction --targets input_examples/reations_gene_to_delete/knockout_reactions_list.txt --faa input_examples/faa_file/e_coli_test.faa --mode double"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af2fd683-a51b-41ad-a57e-db8672137024",
   "metadata": {},
   "source": [
    "# Finish \n",
    "\n",
    "I believe that's it! With this tutorial, you can start building your own GSMR.\n",
    "\n",
    "If you have any issues, please let me know via [GitHub](https://github.com/tamascogustavo/chimera) or contact me via *tamascogustavo@gmail.com*\n",
    "\n",
    "\n",
    "Oh ! **Don't forget to cite us and the others** ✌️"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80cfcdc2-d32a-4ea3-9c9c-b02953643e63",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.3 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
